lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp

```diff
@@ -64,16 +95,121 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
+  auto nextII = std::next(loadInst->getIterator());
+
+  switch (loadInst->getOwnershipQualifier()) {
+  case LoadOwnershipQualifier::Unqualified:
+  case LoadOwnershipQualifier::Trivial:
+    break;
+  case LoadOwnershipQualifier::Take:
+  case LoadOwnershipQualifier::Copy:
+    // Skip ahead if the ownership qualifier isn't handled.
+    //
+    // FIXME: To handle Copy/Take, the code below needs to properly peek through
+    // and destroy borrow scopes.
+    return nextII;
+  }
+  struct ProjInstPair {
+    Projection proj;
+    SingleValueInstruction *svi;
+
+    // When sorting, just look at the projection and ignore the instruction.
+    bool operator<(const ProjInstPair &rhs) const { return proj < rhs.proj; }
+  };
+
+  // Add load projections to a projection list.
+  llvm::SmallVector<ProjInstPair, 8> projections;
+  for (auto *use : getNonDebugUses(loadInst)) {
+    auto *user = use->getUser();
+
+    // If we have any non SEI, TEI instruction, don't do anything here.
+    if (!isa<StructExtractInst>(user) && !isa<TupleExtractInst>(user))
+      return nextII;
+
+    auto extract = cast<SingleValueInstruction>(user);
+    projections.push_back({Projection(extract), extract});
+  }
+  // Avoid removing dead loads from raw SIL. That may weaken definite-init.
+  if (projections.empty() && loadInst->getModule().getStage() == SILStage::Raw)
+    return nextII;
+
+  // Sort the list so projections with the same value decl and tuples with the
+  // same indices will be processed together. This makes it easy to reuse the
+  // load from the first such projection for all subsequent projections on the
+  // same value decl or index.
+  std::sort(projections.begin(), projections.end());
+
+  // Create a new address projection instruction and load instruction for each
+  // unique projection.
+  Projection *lastProj = nullptr;
+  LoadInst *lastNewLoad = nullptr;
+  for (auto &pair : projections) {
+    auto &proj = pair.proj;
+    auto *svi = pair.svi;
+
+    // If this projection is the same as the last projection we processed, just
+    // replace all uses of the projection with the load we created previously.
+    if (lastProj && proj == *lastProj) {
+      LLVM_DEBUG(llvm::dbgs() << "Replacing " << *svi
+                 << "    with " << *lastNewLoad<< "\n");
+      svi->replaceAllUsesWith(lastNewLoad);
+      nextII = killInstruction(svi, nextII, pass);
+      continue;
+    }
+
+    // This is a unique projection. Create the new address projection and load.
+    lastProj = &proj;
+    // Insert new instructions before the original load.
+    SILBuilderWithScope B(loadInst);
+    auto *projInst = proj.createAddressProjection(B, loadInst->getLoc(),
+                                                  loadInst->getOperand())
+                         .get();
+    pass.notifyNewInstruction(projInst);
+
+    lastNewLoad = B.createLoad(loadInst->getLoc(), projInst,
+                               loadInst->getOwnershipQualifier());
```

Comment:
Even though I think that today this is safe (since you are only handling load [trivial], load), this may be incorrect in future cases where we handle [copy], [take]. Specifically if we are shrinking a projection to just use a trivial part of it, then you want to use SILBuilder::createTrivialLoadOr(...). If I am correct, then I am fine if you either use that API or if you add an assert here with a comment saying how this code needs to be updated when this code handles load [copy], load [take].

Reply:
This supports copies in the next commit in the same PR. The SILBuilder utility doesn't help because the client code needs to know how the load will be qualified to generate destroys correctly (the utility looks like it can cause hidden bugs).